src/abstractsparsearrayinterface.jl

```diff
@@ -261,19 +269,18 @@ end
     @interface interface map_all!(f, a_dest, as...)
     return a_dest
   end
-  # First zero out the destination.
-  # TODO: Make this more nuanced, skip when possible, for
-  # example if the sparsity of the destination is a subset of
-  # the sparsity of the sources, i.e.:
-  # ```julia
-  # if eachstoredindex(as...) ∉ eachstoredindex(a_dest)
-  #   zero!(a_dest)
-  # end
-  # ```
-  # This is the safest thing to do in general, for example
-  # if the destination is dense but the sources are sparse.
-  @interface interface zero!(a_dest)
-  @interface interface map_stored!(f, a_dest, as...)
+  # Unalias the inputs from the destination
+  # to make sure the inputs aren't overwritten incorrectly.
+  # See: https://github.com/JuliaLang/julia/blob/v1.11.2/base/broadcast.jl#L935-L948
+  as = map(a -> a_dest === a ? a : Base.unalias(a_dest, a), as)
+  indices_stored = eachstoredindex(as...)
+  if eachstoredindex(a_dest) ⊈ indices_stored
+    # If not all indices being mapped over are stored in the destination,
+    # zero out the destination. An extreme example of this is when
+    # the sources are sparse but the destination is dense.
+    @interface interface zero!(a_dest)
+  end
```

Comment:
I'm not entirely convinced the branch is worth it. If I'm not mistaken, checking the subset is ``O(nnz)``, which means that checking that not all indices are being mapped over takes a considerable amount of time. I would think that a default implementation should simply zero out the resulting memory unconditionally, and a specific type might have a different implementation if they have more information. It's a bit hard to reason about a good default implementation for "any sparse storage", so I would prefer conservative fallbacks that are guaranteed to work.
If you think that this check is worth it though, perhaps it could be worth it to try to refactor that logic into a separate function, which can be specialized to alter the behavior